idx/memory/tag_query.go

```diff
@@ -171,247 +239,812 @@ func NewTagQuery(expressions []string, from int64) (TagQuery, error) {
+	if len(q.tagPrefix) <= 4 && q.tagPrefix == "name"[:len(q.tagPrefix)] {
+		return true
+	}
+
+	for _, tag := range def.Tags {
+		if len(tag) < len(q.tagPrefix) {
+			continue
+		}
+
+		if tag[:len(q.tagPrefix)] == q.tagPrefix {
+			return true
+		}
 	}
 
-	for id := range resultSet {
+	return false
+}
+
+// testByEqual filters a given metric by the defined "=" expressions
+func (q *TagQuery) testByEqual(id idx.MetricID, exprs []kv, not bool) bool {
+	for _, e := range exprs {
+		indexIds := q.index[e.key][e.value]
+
+		// shortcut if key=value combo does not exist at all
+		if len(indexIds) == 0 {
+			return not
+		}
+
+		if _, ok := indexIds[id]; ok {
+			if not {
+				return false
+			}
+		} else {
+			if !not {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// filterIdsFromChan takes a channel of metric ids and runs them through the
+// required tests to decide whether a metric should be part of the final
+// result set or not
+// it returns the final result set via the given resCh parameter
+func (q *TagQuery) filterIdsFromChan(idCh, resCh chan idx.MetricID, completeCh chan struct{}) {
+	for id := range idCh {
 		var def *idx.Archive
 		var ok bool
-		if def, ok = byId[id.String()]; !ok {
+
+		if def, ok = q.byId[id.String()]; !ok {
 			// should never happen because every ID in the tag index
 			// must be present in the byId lookup table
 			corruptIndex.Inc()
 			log.Error(3, "memory-idx: ID %q is in tag index but not in the byId lookup table", id.String())
-			delete(resultSet, id)
 			continue
 		}
 
-		if def.LastUpdate < q.from {
-			delete(resultSet, id)
+		if q.testByAllExpressions(id, def, false) {
+			resCh <- id
 		}
 	}
-}
 
-func (q *TagQuery) Run(index TagIndex, byId map[string]*idx.Archive) TagIDs {
-	var resultSet TagIDs
+	q.wg.Done()
+	completeCh <- struct{}{}
+}
 
+// sortByCost tries to estimate the cost of different expressions and sort them
+// in increasing order
+// this is to reduce the result set cheaply and only apply expensive tests to an
+// already reduced set of results
+func (q *TagQuery) sortByCost() {
 	for i := range q.equal {
-		q.equal[i].cost = uint(len(index[q.equal[i].key][q.equal[i].value]))
+		q.equal[i].cost = uint(len(q.index[q.equal[i].key][q.equal[i].value]))
+	}
+
+	for i := range q.prefix {
+		q.prefix[i].cost = uint(len(q.index[q.prefix[i].key][q.prefix[i].value]))
 	}
 
 	for i := range q.match {
-		q.match[i].cost = uint(len(index[q.match[i].key]))
+		q.match[i].cost = uint(len(q.index[q.match[i].key]))
 	}
 
 	sort.Sort(KvByCost(q.equal))
 	sort.Sort(KvByCost(q.notEqual))
+	sort.Sort(KvByCost(q.prefix))
 	sort.Sort(KvReByCost(q.match))
 	sort.Sort(KvReByCost(q.notMatch))
+}
 
-	if q.startWith == EQUAL {
-		resultSet = q.getInitialByEqual(index, q.equal[0])
-		q.equal = q.equal[1:]
+// Run executes the tag query on the given index and returns a list of ids
+func (q *TagQuery) Run(index TagIndex, byId map[string]*idx.Archive) TagIDs {
```

Comment:
the TagIDs type is confusing. it's not a list but a set. and it's not id's of tags but id's of metrics.

Reply:
does that mean i should change it? I guess i could rename it to something like `IdSet`, but it seems more consistent like this: https://github.com/grafana/metrictank/blob/master/idx/memory/memory.go#L63-L65

Reply:
that, or MetricIDs I suppose. I like IdSet

Reply:
k, renamed it to `IdSet`